apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx

```diff
@@ -28,30 +38,120 @@ class ProgressTableLevelBubble extends React.PureComponent {
     isPaired: PropTypes.bool,
     bubbleSize: PropTypes.oneOf(Object.values(BubbleSize)).isRequired,
     title: PropTypes.string,
-    url: PropTypes.string
+    url: PropTypes.string,
+    useCache: PropTypes.bool
   };
 
   static defaultProps = {
     bubbleSize: BubbleSize.full
   };
 
-  renderContent() {
+  isEnabled() {
+    return !this.props.isDisabled && !this.isLocked();
+  }
+
+  isLocked() {
+    return this.props.levelStatus === LevelStatus.locked;
+  }
+
+  getBubbleShape() {
+    return this.props.isUnplugged
+      ? BubbleShape.pill
+      : this.props.isConcept
+      ? BubbleShape.diamond
+      : BubbleShape.circle;
+  }
+
+  /**
+   * We can't use our usual `progressStyles.hoverStyle` for hover effect here
+   * because we can't use radium in our cached html, so instead we
+   * conditionally add a CSS class to accomplish the same thing.
+   */
+  getClassNames() {
+    return classNames('progress-bubble', {
+      enabled: this.isEnabled()
+    });
+  }
+
+  /**
+   * Determines the simplest key to use for this bubble configuration.
+   */
+  getCacheKey() {
     const {
+      levelId,
       levelStatus,
+      levelKind,
       isUnplugged,
       isBonus,
       isPaired,
       title,
       bubbleSize
     } = this.props;
+
+    let statusString = `status=${levelStatus}`;
+    if (levelKind === LevelKind.assessment) {
+      statusString = `assessment:${statusString}`;
+    }
+
+    if (bubbleSize === BubbleSize.letter) {
+      return `letter:title=${title}&${statusString}`;
+    } else if (isUnplugged) {
+      `unplugged:${statusString}`;
+    }
+
+    const shapeString = `shape=${this.getBubbleShape()}`;
+    const contentString = this.isLocked()
+      ? `locked:`
+      : isPaired
+      ? `paired:`
+      : isBonus
+      ? `bonus:`
+      : title
+      ? `title=${title}`
+      : `levelId=${levelId}`;
+
+    return `${contentString}&${shapeString}&${statusString}`;
+  }
+
+  /**
+   * Retrieve cached html if we have it, otherwise create and cache it.
+   */
+  getOrCreateHtml() {
+    const cacheKey = this.getCacheKey();
+    let bubbleHtml = bubbleHtmlCache[cacheKey];
+    if (!bubbleHtml) {
+      bubbleHtml = this.createHtml();
+      bubbleHtmlCache[cacheKey] = bubbleHtml;
+    }
+    return bubbleHtml;
+  }
+
+  createHtml() {
+    return ReactDOMServer.renderToStaticMarkup(this.createBubbleElement());
+  }
+
+  createBubbleElement() {
+    return React.createElement(
```

Comment:
Mostly curious, why can't we use JSX here?

Reply:
thanks for asking, turns out we can! i took me a few tries to get the syntax right for getting the raw HTML, and at one point i thought i concluded this was necessary. but apparently it's not, so i will update to make this code ever so slightly less obtuse